File: LeetCode/0841. Keys and Rooms/solution.cpp
/**
 * 68 / 68 test cases passed.
 * Runtime: 12 ms
 * Memory Usage: 10.1 MB 
 */
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        int n = rooms.size();
        vector<bool> visit(n);
        queue<int> que;
        for (int i : rooms[0]) {
            que.emplace(i);
        }
        while (!que.empty()) {
            auto sr = que.front(); que.pop();
            visit[sr] = true;
            for (int sc: rooms[sr]) {
                if (!visit[sc]) 
                    que.emplace(sc);
            }
        }
        for (int i = 1; i < n; ++ i) {
            if (!visit[i]) return false;
        }
        return true;
    }
};

File: LeetCode/0841. Keys and Rooms/solution.py
"""
68 / 68 test cases passed.
Runtime: 40 ms
Memory Usage: 15.5 MB
"""
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        n = len(rooms)
        vis = [False] * n
        que = collections.deque(rooms[0])
        while que:
            sr = que.popleft()
            vis[sr] = True
            for sc in rooms[sr]:
                if not vis[sc]:
                    que.append(sc)

        for i in range(1, n):
            if not vis[i]:
                return False
        return True
    